src/MiddlewareRunner.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace React\Http;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use React\Promise;
+use React\Promise\PromiseInterface;
+
+final class MiddlewareRunner
+{
+    /**
+     * @var ResponseInterface
+     */
+    private $defaultResponse;
+
+    /**
+     * @var callable[]
+     */
+    private $middleware = array();
+
+    /**
+     * @param ResponseInterface $response
+     * @param callable[] $middleware
+     */
+    public function __construct(ResponseInterface $response, array $middleware)
+    {
+        $this->defaultResponse = $response;
+        $this->middleware = $middleware;
+    }
+
+    /**
+     * @param ServerRequestInterface $request
+     * @return PromiseInterface<ResponseInterface>
+     */
+    public function __invoke(ServerRequestInterface $request)
+    {
+        if (count($this->middleware) === 0) {
+            return Promise\resolve($this->defaultResponse);
+        }
+
+        $middlewareCollection = $this->middleware;
+        $middleware = array_shift($middlewareCollection);
+
+        $defaultResponse = $this->defaultResponse;
+        $cancel = null;
+        return new Promise\Promise(function ($resolve, $reject) use ($middleware, $request, $middlewareCollection, &$cancel, $defaultResponse) {
+            $cancel = $middleware(
+                $request,
+                new MiddlewareRunner(
+                    $defaultResponse,
+                    $middlewareCollection
+                )
+            );
```

Comment:
I'm okay with this recursive implementation for now, but this is certainly something we would want to look into again in the future :+1:

Reply:
Yup, it works for now and I'll revisit it in the future :+1: